src/ot_ctx.erl

```diff
@@ -0,0 +1,144 @@
+%%%------------------------------------------------------------------------
+%% Copyright 2019, OpenTelemetry Authors
+%% Licensed under the Apache License, Version 2.0 (the "License");
+%% you may not use this file except in compliance with the License.
+%% You may obtain a copy of the License at
+%%
+%% http://www.apache.org/licenses/LICENSE-2.0
+%%
+%% Unless required by applicable law or agreed to in writing, software
+%% distributed under the License is distributed on an "AS IS" BASIS,
+%% WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+%% See the License for the specific language governing permissions and
+%% limitations under the License.
+%%
+%% @doc
+%% @end
+%%%-------------------------------------------------------------------------
+-module(ot_ctx).
+
+-export([set_value/3,
+         get_value/2,
+         with_value/4,
+         with_value/5,
+         remove/2,
+         clear/1,
+         set_current/2,
+         get_current/1,
+         http_extractor/2,
+         http_injector/2,
+         http_extractor/3,
+         http_injector/3]).
+
+-type ctx() :: map().
+-type namespace() :: term().
+-type key() :: term().
+-type value() :: term().
+
+-callback set_value(namespace(), key(), value()) -> ok.
+-callback get_value(namespace(), key()) -> value() | undefined.
+-callback remove(namespace(), key()) -> ok.
+-callback clear(namespace()) -> ok.
+-callback set_current(namespace(), ctx()) -> ok.
+-callback get_current(namespace()) -> ctx().
+
+-export_type([ctx/0,
+              key/0,
+              value/0]).
+
+-spec set_value(namespace(), key(), value()) -> ok.
+set_value(Namespace, Key, Value) ->
+    set_value(opentelemetry:get_context_manager(), Namespace, Key, Value).
+
+-spec set_value(module(), namespace(), key(), value()) -> ok.
+set_value(CtxModule, Namespace, Key, Value) ->
+    CtxModule:set_value(Namespace, Key, Value).
+
+-spec get_value(namespace(), key()) -> value().
+get_value(Namespace, Key) ->
+    get_value(opentelemetry:get_context_manager(), Namespace, Key).
+
+-spec get_value(module(), namespace(), key()) -> value().
+get_value(Module, Namespace, Key) ->
+    Module:get_value(Namespace, Key).
+
+-spec with_value(namespace(), key(), value(), fun()) -> ok.
+with_value(Namespace, Key, Value, Fun) ->
+    with_value(opentelemetry:get_context_manager(), Namespace, Key, Value, Fun).
+
+-spec with_value(module(), namespace(), key(), value(), fun()) -> ok.
+with_value(CtxModule, Namespace, Key, Value, Fun) ->
+    PrevCtx = get_current(CtxModule, Namespace),
+    CtxModule:set_current(PrevCtx, Namespace, PrevCtx#{Key => Value}),
+    try
+        Fun()
+    after
+        set_current(CtxModule, Namespace, PrevCtx)
+    end.
```

Comment:
I guess this usage implies that the current values for the context can't be global, and only process-global at best? Otherwise running two of these at the same time may break things, right? Might be worth pointing out in a comment?

Reply:
Yes, context is tied to a process. I can note that.